### PR TITLE
src: Add WebRTC development frontend

### PR DESCRIPTION
**Pull request by joaoantoniocardoso**
I'll soon merge both interfaces to be easier to deal with the build and HTTP server

note: this will *not* build without the backend merged

### CODE REVIEW

File: src/stream/webrtc/frontend/src/consumer.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  public removeAllSessions(): Result<void> {
+    this.sessions.forEach((session: Session): void => session.end());
+    this.sessions.clear();
+    console.debug(
+      `All sessions from consumer ${
+        this.id
+      } removed. Current sessions: ${JSON.stringify(this.sessions, null, 4)}`
+    );
+    return Result.ok(undefined as void);
```
Comment (patrickelectric) [inline src/stream/webrtc/frontend/src/consumer.ts:59]: Is said that we can't return an empty ok

File: src/stream/webrtc/frontend/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { createApp } from "vue";
+import App from "./App.vue";
+
+import "./assets/main.css";
```
Comment (patrickelectric) [inline src/stream/webrtc/frontend/src/main.ts:4]: avoid relative imports, prefer '@'

File: src/stream/webrtc/frontend/src/manager.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        null,
+        4
+      )}`
+    );
+    return Result.ok(undefined as void);
+  }
+
+  public removeAllConsumers(): Result<void> {
+    this.consumers.forEach((consumer: Consumer): void => consumer.end());
+    this.consumers.clear();
```
Comment (patrickelectric) [inline src/stream/webrtc/frontend/src/manager.ts:81]: "garbage collector is a good idea"

**Review by patrickelectric — CHANGES_REQUESTED** [review]
`package-lock.json` is not necessary if we are using yarn as our package manager

File: build.rs [hunk 4]
```diff
@@ -23,4 +35,69 @@ fn main() {
...
+    }
+    let bindings_file_path = output_dir.join(Path::new("signalling_protocol.d.ts"));
+    let mut bindings_file = fs::File::create(bindings_file_path).unwrap();
+    bindings_file.write_all(bindings.as_bytes()).unwrap();
+}
+
+fn build_with_npm() {
+    // Note that as we are not waching all files, sometimes we'd need to force this build
+    println!("cargo:rerun-if-changed=./src/stream/webrtc/frontend/index.html");
+    println!("cargo:rerun-if-changed=./src/stream/webrtc/frontend/package-lock.json");
```
Comment (patrickelectric) [inline build.rs:88]: use yarn lock over package-lock since we are using yarn

File: build.rs [hunk 5]
```diff
@@ -23,4 +35,69 @@ fn main() {
...
         "src/html/vue.js",
     );
+
+    // set SKIP_BINDINGS=1 to skip typescript bindings generation
+    if std::env::var("SKIP_BINDINGS").is_err() {
+        generate_typescript_bindings();
+    }
+
+    // set SKIP_NPM=1 to skip NPM build
+    if std::env::var("SKIP_NPM").is_err() {
```
Comment (patrickelectric) [inline build.rs:45]: YARN

File: build.rs [hunk 6]
```diff
@@ -23,4 +35,69 @@ fn main() {
...
+    let output_dir = Path::new("./src/stream/webrtc/frontend/bindings/");
+    if !output_dir.exists() {
+        std::fs::create_dir_all(output_dir).unwrap();
+    }
+    let bindings_file_path = output_dir.join(Path::new("signalling_protocol.d.ts"));
+    let mut bindings_file = fs::File::create(bindings_file_path).unwrap();
+    bindings_file.write_all(bindings.as_bytes()).unwrap();
+}
+
+fn build_with_npm() {
```
Comment (patrickelectric) [inline build.rs:85]: yarn

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: src/stream/webrtc/frontend/src/App.vue [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+const ip = window.location.hostname.toString();
+// const ip = "blueos.local";
+
+// eslint-disable-next-line no-undef
+const rtc_configuration: RTCConfiguration = {
+  bundlePolicy: "max-bundle",
+  iceServers: [
+    {
+      urls: "turn:0.0.0.0:3478",
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/webrtc/frontend/src/App.vue:23]: great catch, it should

**Review by patrickelectric — APPROVED** [review]
Lets continue working over it

**Pull request merged**